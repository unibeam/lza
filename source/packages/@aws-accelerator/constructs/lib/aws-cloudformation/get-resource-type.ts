/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { NagSuppressions } from 'cdk-nag';

const path = require('path');

/**
 * Get the ResourceType from a CloudFormation Stack by supplying
 * the logicalId
 */
export interface GetCloudFormationResourceTypeProps {
  readonly stackName: string;
  readonly logicalResourceId: string;
  readonly partition: string;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly cloudwatchKmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

export class GetCloudFormationResourceType extends Construct {
  public readonly id: string;

  constructor(scope: Construct, id: string, props: GetCloudFormationResourceTypeProps) {
    super(scope, id);

    const RESOURCE_TYPE = 'Custom::GetCloudFormationResourceType';

    const cloudformationPolicy = new cdk.aws_iam.PolicyStatement({
      sid: 'cloudformation',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['cloudformation:DescribeStackResource'],
      resources: [`arn:${props.partition}:cloudformation:*:${cdk.Stack.of(this).account}`],
    });

    const lambdaFunction = new cdk.aws_lambda.Function(this, 'GetCloudFormationResourceTypeFunction', {
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, 'get-resource-type/dist')),
      runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      timeout: cdk.Duration.minutes(3),
      description: 'Get CloudFormation Resources from Stack by LogicalResourceId',
      initialPolicy: [cloudformationPolicy],
    });

    const logGroup = new cdk.aws_logs.LogGroup(this, `${lambdaFunction.node.id}LogGroup`, {
      logGroupName: `/aws/lambda/${lambdaFunction.functionName}`,
      retention: props.logRetentionInDays,
      encryptionKey: props.cloudwatchKmsKey,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    const provider = new cdk.custom_resources.Provider(this, 'GetCloudFormationResourceTypeProvider', {
      onEventHandler: lambdaFunction,
    });

    const resource = new cdk.CustomResource(this, 'GetCloudFormationResourceTypeResource', {
      resourceType: RESOURCE_TYPE,
      serviceToken: provider.serviceToken,
      properties: {
        stackName: props.stackName,
        logicalResourceId: props.logicalResourceId,
      },
    });

    NagSuppressions.addResourceSuppressions(
      lambdaFunction,
      [
        {
          id: 'AwsSolutions-IAM4',
          reason: 'Lambda managed policy',
        },
        {
          id: 'AwsSolutions-IAM5',
          reason: 'Lambda managed policy',
        },
      ],
      true,
    );

    NagSuppressions.addResourceSuppressions(
      provider,
      [
        {
          id: 'AwsSolutions-IAM4',
          reason: 'CDK generated provider',
        },
        {
          id: 'AwsSolutions-IAM5',
          reason: 'CDK generated provider',
        },
      ],
      true,
    );

    // Ensure that the LogGroup is created by Cloudformation prior to Lambda execution
    resource.node.addDependency(logGroup);
    this.id = resource.ref;
  }
}
