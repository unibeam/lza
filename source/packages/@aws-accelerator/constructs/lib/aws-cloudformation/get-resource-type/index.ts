/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
/**
 * get-resource-type - lambda handler
 *
 * @param event
 * @returns
 */

import {
  CloudFormationClient,
  DescribeStackResourceCommand,
  DescribeStackResourceCommandOutput,
} from '@aws-sdk/client-cloudformation';
import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';

export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string | undefined;
      Status: string;
    }
  | undefined
> {
  const stackName = event.ResourceProperties['stackName'];
  const logicalResourceId = event.ResourceProperties['logicalResourceId'];
  const solutionId = process.env['SOLUTION_ID'];
  let resourceDetails: DescribeStackResourceCommandOutput | undefined = undefined;
  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      const cloudformationClient = new CloudFormationClient({ customUserAgent: solutionId });
      try {
        resourceDetails = await throttlingBackOff(() =>
          cloudformationClient.send(
            new DescribeStackResourceCommand({ StackName: stackName, LogicalResourceId: logicalResourceId }),
          ),
        );
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } catch (e: any) {
        if (e.name === 'ValidationError') {
          return {
            PhysicalResourceId: undefined,
            Status: 'SUCCESS',
          };
        } else {
          console.log(`Error: ${JSON.stringify(e)}`);
          throw new Error(`Error retrieving CloudFormation stack named ${stackName}`);
        }
      }
      if (resourceDetails && resourceDetails.StackResourceDetail?.LogicalResourceId) {
        return {
          PhysicalResourceId: resourceDetails.StackResourceDetail?.ResourceType,
          Status: 'SUCCESS',
        };
      }
      return {
        PhysicalResourceId: undefined,
        Status: 'SUCCESS',
      };
    case 'Delete':
      // Do Nothing
      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }
}
