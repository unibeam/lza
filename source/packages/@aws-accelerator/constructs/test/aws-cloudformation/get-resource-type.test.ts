/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { GetCloudFormationResourceType } from '../../lib/aws-cloudformation/get-resource-type';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(GetCloudFormationResourceType): ';

const app = new cdk.App();

// Create stack for native Cfn construct
const env = { account: '333333333333', region: 'us-east-1' };
const stack = new cdk.Stack(app, 'Stack', { env: env });

new GetCloudFormationResourceType(stack, 'TestGetCloudFormationResourceType', {
  stackName: 'AWSAccelerator-TestStack-us-east-1',
  logicalResourceId: 'TestResource',
  logRetentionInDays: 3653,
  cloudwatchKmsKey: new cdk.aws_kms.Key(stack, 'CustomKey', {}),
  partition: 'aws',
});

/**
 * ConfigServiceTags construct test
 */
describe('GetCloudFormationResourceType', () => {
  snapShotTest(testNamePrefix, stack);
});
