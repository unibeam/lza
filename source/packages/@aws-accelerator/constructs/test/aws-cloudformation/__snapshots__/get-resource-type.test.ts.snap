// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetCloudFormationResourceType Construct(GetCloudFormationResourceType):  Snapshot Test 1`] = `
{
  "Resources": {
    "CustomKey1E6D0D07": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::333333333333:root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionA13B89C3": {
      "DependsOn": [
        "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionServiceRoleDefaultPolicy7A52BF9E",
        "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionServiceRole450819BD",
      ],
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM4",
              "reason": "Lambda managed policy",
            },
            {
              "id": "AwsSolutions-IAM5",
              "reason": "Lambda managed policy",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-333333333333-us-east-1",
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "Get CloudFormation Resources from Stack by LogicalResourceId",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionServiceRole450819BD",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 180,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionLogGroup602C5B86": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "CustomKey1E6D0D07",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionA13B89C3",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionServiceRole450819BD": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM4",
              "reason": "Lambda managed policy",
            },
            {
              "id": "AwsSolutions-IAM5",
              "reason": "Lambda managed policy",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionServiceRoleDefaultPolicy7A52BF9E": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM4",
              "reason": "Lambda managed policy",
            },
            {
              "id": "AwsSolutions-IAM5",
              "reason": "Lambda managed policy",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "cloudformation:DescribeStackResource",
              "Effect": "Allow",
              "Resource": "arn:aws:cloudformation:*:333333333333",
              "Sid": "cloudformation",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionServiceRoleDefaultPolicy7A52BF9E",
        "Roles": [
          {
            "Ref": "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionServiceRole450819BD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeProviderframeworkonEvent83E12901": {
      "DependsOn": [
        "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeProviderframeworkonEventServiceRoleDefaultPolicyB5AC412E",
        "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeProviderframeworkonEventServiceRoleFB2D11F4",
      ],
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM4",
              "reason": "CDK generated provider",
            },
            {
              "id": "AwsSolutions-IAM5",
              "reason": "CDK generated provider",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-333333333333-us-east-1",
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (Stack/TestGetCloudFormationResourceType/GetCloudFormationResourceTypeProvider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionA13B89C3",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeProviderframeworkonEventServiceRoleFB2D11F4",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeProviderframeworkonEventServiceRoleDefaultPolicyB5AC412E": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM4",
              "reason": "CDK generated provider",
            },
            {
              "id": "AwsSolutions-IAM5",
              "reason": "CDK generated provider",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionA13B89C3",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionA13B89C3",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeProviderframeworkonEventServiceRoleDefaultPolicyB5AC412E",
        "Roles": [
          {
            "Ref": "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeProviderframeworkonEventServiceRoleFB2D11F4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeProviderframeworkonEventServiceRoleFB2D11F4": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-IAM4",
              "reason": "CDK generated provider",
            },
            {
              "id": "AwsSolutions-IAM5",
              "reason": "CDK generated provider",
            },
          ],
        },
      },
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeResource41628DFB": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeFunctionLogGroup602C5B86",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "TestGetCloudFormationResourceTypeGetCloudFormationResourceTypeProviderframeworkonEvent83E12901",
            "Arn",
          ],
        },
        "logicalResourceId": "TestResource",
        "stackName": "AWSAccelerator-TestStack-us-east-1",
      },
      "Type": "Custom::GetCloudFormationResourceType",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
